swaylock/main.c

```diff
@@ -621,78 +614,212 @@ int main(int argc, char **argv) {
 			break;
 		case LO_TEXT_CLEAR_COLOR:
-			state.args.colors.text.cleared = parse_color(optarg);
+			state->args.colors.text.cleared = parse_color(optarg);
 			break;
 		case LO_TEXT_VER_COLOR:
-			state.args.colors.text.verifying = parse_color(optarg);
+			state->args.colors.text.verifying = parse_color(optarg);
 			break;
 		case LO_TEXT_WRONG_COLOR:
-			state.args.colors.text.wrong = parse_color(optarg);
+			state->args.colors.text.wrong = parse_color(optarg);
 			break;
 		default:
 			fprintf(stderr, "%s", usage);
 			return 1;
 		}
 	}
 
+	return 0;
+}
+
+static bool file_exists(const char *path) {
+	return path && access(path, R_OK) != -1;
+}
+
+static char *get_config_path(void) {
+	static const char *config_paths[] = {
+		"$HOME/.swaylock/config",
+		"$XDG_CONFIG_HOME/swaylock/config",
+		SYSCONFDIR "/swaylock/config",
+	};
+
+	if (!getenv("XDG_CONFIG_HOME")) {
+		char *home = getenv("HOME");
+		char *config_home = malloc(strlen(home) + strlen("/.config") + 1);
+		if (!config_home) {
+			wlr_log(WLR_ERROR, "Unable to allocate $HOME/.config");
+		} else {
+			strcpy(config_home, home);
+			strcat(config_home, "/.config");
+			setenv("XDG_CONFIG_HOME", config_home, 1);
+			wlr_log(WLR_DEBUG, "Set XDG_CONFIG_HOME to %s", config_home);
+			free(config_home);
+		}
+	}
+
+	wordexp_t p;
+	char *path;
+	for (int i = 0; i < (int)(sizeof(config_paths) / sizeof(char *)); ++i) {
+		if (wordexp(config_paths[i], &p, 0) == 0) {
+			path = strdup(p.we_wordv[0]);
+			wordfree(&p);
+			if (file_exists(path)) {
+				return path;
+			}
+			free(path);
+		}
+	}
+
+	return NULL;
+}
+
+static int load_config(char *path, struct swaylock_state *state,
+		enum line_mode *line_mode) {
+	FILE *config = fopen(path, "r");
+	if (!config) {
+		wlr_log(WLR_ERROR, "Failed to read config. Running without it.");
+		return 0;
+	}
+	char *line;
+	int line_number = 0;
+	while (!feof(config)) {
+		line = read_line(config);
+		if (!line) {
+			continue;
+		}
+
+		line_number++;
+		if (line[0] == '#') {
+			free(line);
+			continue;
+		}
+		if (strlen(line) == 0) {
+			free(line);
+			continue;
+		}
+
+		wlr_log(WLR_DEBUG, "Config Line #%d: %s", line_number, line);
+		char flag[strlen(line) + 3];
+		sprintf(flag, "--%s", line);
+		char *argv[] = {"swaylock", flag};
+		int result = parse_options(2, argv, state, line_mode);
+		if (result != 0) {
+			free(line);
+			fclose(config);
+			return result;
+		}
+		free(line);
+	}
+	fclose(config);
+	return 0;
+}
+
+static struct swaylock_state state;
+
+int main(int argc, char **argv) {
+	enum line_mode line_mode = LM_LINE;
+	state.args = (struct swaylock_args){
+		.mode = BACKGROUND_MODE_SOLID_COLOR,
+		.font = strdup("sans-serif"),
+		.radius = 50,
+		.thickness = 10,
+		.ignore_empty = false,
+		.show_indicator = true,
+	};
+	wl_list_init(&state.images);
+	set_default_colors(&state.args.colors);
+
+	wlr_log_init(WLR_DEBUG, NULL);
+
+	char *config_path = NULL;
+	for (int i = 0; i < argc; i++) {
```

Comment:
Instead, `parse_options` could be a more simple `parse_option` function which parses a single option key/value.

Reply:
That increases the complexity of command line arguments since `--color <rrggbb[aa]>` (which is two arguments) could be given instead of `--color=<rrggbb[aa]>`. I guess I could make the `=` mandatory for command line arguments or detect whether there is a `=` in the argument.

Reply:
Eh, doesn't getopt handle that for you? We should stick to getopt's behaviour.

Reply:
(I was thinking of something like `parse_option(state, key, value)`)

Reply:
After discussing on IRC, we decided to keep it as is. @emersion's complaint was with the manual `-C, --config` handling. Since the config file needs to be loaded first and then have command line overrides, we agreed to keep it as is.